### PR TITLE
feat: Implement Zstandard compression codec for gRPC

### LINKED ISSUE #13265: HIP-1081 Support custom zstd compression gRPC streaming
### Problem

PBJ added custom zstd compression gPRC streaming that'll reduce data size over the wire between CN and BN. MN uses grpc-netty and needs to implement equivalent custom zstd compression support for the streaming from BN to MN.

### Solution

TBA

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by karthik120710**
- Added ZstdCodec class for compressing and decompressing data using Zstandard.
- Updated ManagedChannelBuilderProviderImpl to utilize ZstdCodec for gRPC compression and decompression.
- Changed zstd-jni dependency from runtimeOnly to implementation in build.gradle.kts.
- Added unit tests for ZstdCodec to ensure correct functionality.

**Description**:
feat: Implement Zstandard compression codec for gRPC (#13265)
- Add a ZstdCodec implementing gRPC's Codec interface to enable Zstandard (zstd) compression for gRPC communication with block nodes. The codec is registered with both the compressor and decompressor registries in ManagedChannelBuilderProviderImpl.

 Changes:
- New ZstdCodec class wrapping ZstdInputStream/ZstdOutputStream for gRPC compression/decompression
- Register Zstd codec on the gRPC ManagedChannel in ManagedChannelBuilderProviderImpl
- Promote zstd-jni from runtimeOnly to implementation scope in build config
- Add unit tests verifying roundtrip compress/decompress for empty, small, and large payloads
- #13265

### CODE REVIEW

**Comment by karthik120710** [comment]
hi @nickeynikolovv i wrote zstd compression support for grpc streaming, can u review my pr.

File: importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+package org.hiero.mirror.importer.downloader.block;
+
+import com.github.luben.zstd.ZstdInputStream;
+import com.github.luben.zstd.ZstdOutputStream;
+import io.grpc.Codec;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+class ZstdCodec implements Codec {
```
Comment (steven-sheehy) [inline importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java]:
```suggestion
final class ZstdCodec implements Codec {
```

Comment (steven-sheehy) [inline importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java]: Since this doesn't have per stream state we should be able to make this a bean and inject it instead of using new operator. Add `@Named` and inject using `@RequiredArgsConstructor`.

File: importer/src/main/java/org/hiero/mirror/importer/downloader/block/ManagedChannelBuilderProviderImpl.java [hunk 2]
```diff
@@ -10,13 +12,16 @@ final class ManagedChannelBuilderProviderImpl implements ManagedChannelBuilderPr
...
+        final var zstdCodec = new ZstdCodec();
         final var builder = ManagedChannelBuilder.forAddress(host, port);
         if (useTls) {
             builder.useTransportSecurity();
         } else {
             builder.usePlaintext();
         }
 
+        builder.decompressorRegistry(DecompressorRegistry.getDefaultInstance().with(zstdCodec, true));
+        builder.compressorRegistry(CompressorRegistry.getDefaultInstance().with(zstdCodec));
```
Comment (steven-sheehy) [inline importer/src/main/java/org/hiero/mirror/importer/downloader/block/ManagedChannelBuilderProviderImpl.java]: Does this do content negotiation such that the stream can still work with block nodes that don't support zstd? Or does this now force the stream to always use zstd?

File: importer/src/test/java/org/hiero/mirror/importer/downloader/block/ZstdCodecTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ZstdCodecTest {
```
Comment (steven-sheehy) [inline importer/src/test/java/org/hiero/mirror/importer/downloader/block/ZstdCodecTest.java]: final

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Comment by karthik120710** [comment]
hi @steven-sheehy can u review my pr.

File: importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java [hunk 4]
```diff
@@ -5,11 +5,15 @@
...
 import io.grpc.Codec;
+import jakarta.inject.Named;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import lombok.RequiredArgsConstructor;
 
-class ZstdCodec implements Codec {
+@Named
+@RequiredArgsConstructor
```
Comment (steven-sheehy) [inline importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java]: This annotation is redundant since no fields.

File: importer/src/main/java/org/hiero/mirror/importer/downloader/block/ManagedChannelBuilderProviderImpl.java (hunk 2 above)
Comment (steven-sheehy) [reply importer/src/main/java/org/hiero/mirror/importer/downloader/block/ManagedChannelBuilderProviderImpl.java]: You didn't answer this question

Comment (karthik120710) [reply importer/src/main/java/org/hiero/mirror/importer/downloader/block/ManagedChannelBuilderProviderImpl.java]: @steven-sheehy No, it doesn't force zstd. DecompressorRegistry.with(zstdCodec, true) just adds zstd to the grpc-accept-encoding header, advertising that the client can decompress it. The server decides whether to use it. Block nodes that don't support zstd simply send uncompressed responses — gRPC handles this automatically via content negotiation. The stream works with all nodes either way.

File: importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java (hunk 4 above)
Comment (karthik120710) [reply importer/src/main/java/org/hiero/mirror/importer/downloader/block/ZstdCodec.java]: i will remove RequiredArgsConstructor

**Comment by karthik120710** [comment]
hi @steven-sheehy i have removed the @RequiredArgsConstructor anotation can u review my pr .

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
LGTM. Thanks for your contribution!
